packages/core/src/api/apis/implementations/lib/AuthConnector/DefaultAuthConnector.ts

```diff
@@ -0,0 +1,195 @@
+  private readonly apiOrigin: string;
+  private readonly basePath: string;
+  private readonly environment: string;
+  private readonly provider: AuthProvider & { id: string };
+  private readonly authRequester: AuthRequester<AuthSession>;
+  private readonly sessionTransform: (response: any) => Promise<AuthSession>;
+
+  constructor(options: Options<AuthSession>) {
+    const {
+      apiOrigin = window.location.origin,
+      basePath = DEFAULT_BASE_PATH,
+      environment,
+      provider,
+      joinScopes = defaultJoinScopes,
+      oauthRequestApi,
+      sessionTransform = id => id,
+    } = options;
+
+    this.authRequester = oauthRequestApi.createAuthRequester({
+      provider,
+      onAuthRequest: scopes => this.showPopup(joinScopes(scopes)),
+    });
+
+    this.apiOrigin = apiOrigin;
+    this.basePath = basePath;
+    this.environment = environment;
+    this.provider = provider;
+    this.sessionTransform = sessionTransform;
+  }
+
+  async createSession(scopes: Set<string>): Promise<AuthSession> {
+    return this.authRequester(scopes);
+  }
+
+  async refreshSession(): Promise<any> {
+    const res = await fetch(this.buildUrl('/token', { optional: true }), {
+      headers: {
+        'x-requested-with': 'XMLHttpRequest',
+      },
+      credentials: 'include',
+    }).catch(error => {
+      throw new Error(`Auth refresh request failed, ${error}`);
+    });
+
+    if (!res.ok) {
+      const error: any = new Error(
+        `Auth refresh request failed with status ${res.statusText}`,
+      );
+      error.status = res.status;
+      throw error;
+    }
+
+    const authInfo = await res.json();
+
+    if (authInfo.error) {
+      const error = new Error(authInfo.error.message);
+      if (authInfo.error.name) {
+        error.name = authInfo.error.name;
+      }
+      throw error;
+    }
+    return await this.sessionTransform(authInfo);
+  }
+
+  async removeSession(): Promise<void> {
+    const res = await fetch(this.buildUrl('/logout'), {
+      method: 'POST',
+      headers: {
+        'x-requested-with': 'XMLHttpRequest',
+      },
+      credentials: 'include',
+    });
+
+    if (!res.ok) {
+      throw new Error(`Logout request failed with status ${res.status}`);
+    }
+  }
+
+  private async showPopup(scope: string): Promise<AuthSession> {
+    const popupUrl = this.buildUrl('/start', { scope });
+
+    const payload = await showLoginPopup({
+      url: popupUrl,
+      name: `${this.provider.title} Login`,
+      origin: this.apiOrigin,
+      width: 450,
+      height: 730,
+    });
+
+    return await this.sessionTransform(payload);
+  }
+
+  private buildUrl(
+    path: string,
+    query?: { [key: string]: string | boolean | undefined },
+  ): string {
+    const queryString = this.buildQueryString({
+      ...query,
+      env: this.environment,
+    });
+
+    return `${this.apiOrigin}${this.basePath}${this.provider.id}${path}${queryString}`;
+  }
+
+  private buildQueryString(query?: {
```

Comment:
isn't this what `qs` is great at

Reply:
Iirc I started out with `URLSearchParams` but got rid of that because we needed boolean params to appear without value. Don't think I tried that with qs though tbh.